libraries/platforms/dromajo-vcs/bsg_manycore_simulator.cpp

```diff
@@ -0,0 +1,696 @@
+  bool is_empty;
+
+  do {
+    // Read the FIFO head pointer for all 32-bit FIFOs
+    dromajo_to_mc_packet.words[0] = host_to_mc_req_fifo->fifo[0].front();
+    dromajo_to_mc_packet.words[1] = host_to_mc_req_fifo->fifo[1].front();
+    dromajo_to_mc_packet.words[2] = host_to_mc_req_fifo->fifo[2].front();
+    dromajo_to_mc_packet.words[3] = host_to_mc_req_fifo->fifo[3].front();
+
+    // Intercept packets that are for the host and generate appropriate responses
+    // TODO: Currently, these packets don't go over the network. In the real system (with BP), some or all of these may be required to go
+    // over the network. The platform implementation assumes the host packets do not go over the network. In some cases (like fences) this
+    // might not be desirable. Make sure to keep the platform and the system host in sync on what is actually being simulated.
+    if ((dromajo_to_mc_packet.request.x_dst == HOST_X_COORD) && (dromajo_to_mc_packet.request.y_dst == HOST_Y_COORD)) {
+      err = dromajo_transmit_host_packet(&dromajo_to_mc_packet);
+      // This serves to return the error from the host function
+      // It is also used to communicate the success code for successful program termination but this code is
+      // not HB_MC_SUCCESS (maps to 0) but a positive number
+      return err;
+    }
+    else {
+      pkt = reinterpret_cast<__m128i *>(&dromajo_to_mc_packet);
+
+      // Allows the DPI interface to track response FIFO capacity
+      bool expect_response =
+        (dromajo_to_mc_packet.request.op_v2 != HB_MC_PACKET_OP_REMOTE_STORE) &&
+        (dromajo_to_mc_packet.request.op_v2 != HB_MC_PACKET_OP_REMOTE_SW) &&
+        (dromajo_to_mc_packet.request.op_v2 != HB_MC_PACKET_OP_CACHE_OP);
+
+      // Attempt packet transmission
+      // Since we trigger a call to the transmit FIFOs only when all the 32-bit Dromajo
+      // FIFOs are full (i.e. a single 128-bit packet is ready), we need to wait until
+      // the DPI FIFOs are ready to receive before advancing to the next operation.
+      // This can prevent filling up of the FIFOs. However, not doing this can help in
+      // identifying situations that might create backpressure in actual hardware and
+      // provision for it.
+      do {
+        advance_time();
+        err = dpi->tx_req(*pkt, expect_response);
+      } while (err != BSG_NONSYNTH_DPI_SUCCESS  &&
+           (err == BSG_NONSYNTH_DPI_NO_CREDITS  ||
+            err == BSG_NONSYNTH_DPI_NO_CAPACITY ||
+            err == BSG_NONSYNTH_DPI_NOT_WINDOW  ||
+            err == BSG_NONSYNTH_DPI_BUSY        ||
+            err == BSG_NONSYNTH_DPI_NOT_READY));
+
+      if (err == BSG_NONSYNTH_DPI_SUCCESS) {
+        for (int i = 0;i < 4; i++)
+          host_to_mc_req_fifo->fifo[i].pop();
+      }
+      else {
+        bsg_pr_err("%s: Packet transmission failed\n", __func__);
+        return HB_MC_FAIL;
+      }
+    }
+    is_empty = mc_is_fifo_empty(type);
+  } while (!is_empty);
```

Comment:
Like above, I would make this transmit_device_packet. Then it's clear it's host OR device.


But it's much easier to understand now.